citar.el

```diff
@@ -560,124 +584,136 @@ HISTORY is the `completing-read' history argument."
+            (mapcan
+             (pcase-lambda (`(,cat . ,cands))
+               (if (not cat)
+                   cands
+                 (mapcar (lambda (cand) (propertize cand 'multi-category (cons cat cand))) cands)))
+             sources)))))
 
 (defun citar--annotate-note (candidate)
   "Annotate note CANDIDATE."
-  (when-let* ((notep (get-text-property 0 'notep candidate))
+  (when-let* ((notep (get-text-property 0 'citar--note candidate))
               (annotate (citar--get-notes-config :annotate)))
     (funcall
      annotate
      (substring-no-properties
       (cdr (get-text-property 0 'multi-category candidate))))))
 
-(cl-defun citar--select-resource (keys &optional &key files notes links)
+(cl-defun citar--select-resource (keys &key files notes links (always-prompt t))
   ;; FIX the arg list above is not smart
   "Select related FILES, NOTES, or LINKS resource for KEYS."
-  (if-let ((resources
-            (citar--get-resource-candidates
-             keys :files files :notes notes :links links)))
+  (when-let* ((resources
+               (citar--get-resource-candidates
+                keys :files files :notes notes :links links))
+              (candidates (cdr resources)))
+    (if (and (not always-prompt) (null (cdr candidates)))
+        (car candidates)
       (completing-read
        "Select resource: "
        (lambda (string predicate action)
          (if (eq action 'metadata)
              `(metadata
-               (group-function . citar--select-group-related-resources)
-               (annotation-function . citar--annotate-note)
-               (category . multi-category))
-           (complete-with-action action resources string predicate))))))
+               (group-function . ,#'citar--select-group-related-resources)
+               (annotation-function . ,#'citar--annotate-note)
+               ,@(when-let ((cat (car resources))) `((category . ,cat))))
+           (complete-with-action action candidates string predicate)))))))
 
 (defun citar--select-group-related-resources (resource transform)
   "Group RESOURCE by type or TRANSFORM."
-  (if transform
-      (if (file-regular-p resource)
-          (file-name-nondirectory resource)
-        resource)
-    (let ((cat (car (get-text-property 0 'multi-category resource)))
-          (notep (get-text-property 0 'notep resource)))
-      ;; If note, assign to note group; otherwise use completion category.
-      (if notep (citar--get-notes-config :name)
-        (pcase cat
-          ('file "Library Files")
-          ('url "Links"))))))
-
-(cl-defun citar--format-candidates (&key (bibs (citar--bibliographies))
-                                         (entries (citar-cache--entries bibs)))
-  "Format completion candidates for ENTRIES.
-
-BIBS should be a list of `citar-cache--bibliography' objects that
-are the source of ENTRIES. Use the pre-formatted strings in BIBS
-to format candidates.
+  (if (get-text-property 0 'citar--note resource)
```

Comment:
The configurable transform functions are just an side benefit. To make `citar-open-files` and `citar-open-links` wrappers around `citar--select-resource`, I changed `citar--get-resource-candidates` to return the completion category for the list it returns; it only uses `multi-category` if there is more than category of item.

The old group function would only assign groups to the candidates if they had the `multi-category` property. But Vertico (and probably other completion UIs) ignore the transformation if there is no group assigned. This created a bug, where you would see transformed candidates only for `citar-open` (a `multi-category` completion) but not for `citar-open-files`.

While fixing that, I decided to restructure the group function so there are three separate branches, for notes, files, and links. Since each branch explicitly both transforms the resource and assigns it to a group, it's much harder to accidentally do one without the other. I thought the small increase in lines of code was worth it for the clarity.

This naturally raised the question of how to transform notes. In the old code, all candidates that satisfied `file-regular-p` were transformed regardless of whether they were notes or library files. Instead of conflating files and notes, I figured it was cleaner for note sources to explicitly specify the transformation via a `:transform` function.